### ISSUE TITLE
Improve modularity of code

### ISSUE THREAD

**Issue by tfukaza** [tfukaza/harvest#232]
One issue with the way the code is organized currently is that each module frequently makes references to each other. For example, the Storage class gets the current time by referencing `trader.timestamp`, and Trader updates the timestamp by referencing Streamer. This can make the code messy and make the ownership of the attributes unclear. 

My current idea is to bundle common runtime attributes like `timestamp` and `watchlist`, as well as info like current stock positions, into a single class/dictionary. Every module will share a reference to a single instance of this class/dict.

**Comment by shershah010** [comment]
How would the `timestamp` be updated. Obviously it should be done on interval, so would that be in the streamer's main function?

**Comment by tfukaza** [comment]
That was my idea, since all modules share the same _reference_ to a _single_ instance. E.g. if streamer updates the timestamp, the Storage can just say something like `self.stat.timestamp` instead of `self.trader.streamer.timestamp`.

**Comment by shershah010** [comment]
If it is just the `timestamp` then I think that can be part of the streamer and the trader can pass the `self.streamer.timestamp` to the storage class. This way there are no global variables. Additionally we can use the streamer's api to fetch the system time of the the server.
However since I can see this will be a problem since we have both `timestamp` and `timezone` info being passed around.
Would it be beneficial to have a class that tracks system that is agnostic of the streamer and then another one for the statistics. We can of course just make one giant class with all that holds all the global variables. I guess how would this be created? It would either be a new file or in `utils.py`. We also have to work out how to create / update this dictionary / class. I.e. if someone creates a broker and then a trader, the broker needs to create this object and the trader would need to update it.

**Comment by tfukaza** [comment]
That was my thought too. We can make a class like `stats` that gets instantiated by Trader and is passed onto all other modules right before the algo beings running. Any class can update or access values using something like `stat.timestamp = now()`

**Comment by shershah010** [comment]
What if a user only wants to use a broker without the trader? It could be better if the `stats` class is created once globally and each feature can access and modify it.

**Comment by tfukaza** [comment]
In my opinion the broker class shouldn't exist on its own anyways. Personally I'm against making too many things global as controlling its scope can become difficult.

**Comment by shershah010** [comment]
I think that all of out primitives (in this case the brokers and storage, and algo to some extent) should be usable without a class that is built on top of it, i.e. the trader. Like in tensorflow users can use  softmax or use a model with a softmax layer. I guess  the `_base` broker can optionally receive a `stats` class from the trader and if not provided then it creates its own. This would make the code easier to test as less features will rely on the trader class. Also it gives users the ability to work with a broker of their choice without needing to be locked into the trader's structure.
